### PR TITLE
fix(skore): Use display API for coefficients display

### PR DESCRIPTION
**Pull request by glemaitre**
Follow-up to #2043 
PRs needed to be merged before:
- #2181
- #2182

This PR intends to follow more strictly our `Display` API.

- expose a `_compute_data_for_display` in the `Display` class
- store a long format dataframe containing all the information required independently of the type of report
- use `.frame` to drop the unnecessary columns depending of the type of report
- leverage the `.frame` long format to directly use seaborn out of the box plotting API
- add the `subplots_by` API to show on different subplot

### CODE REVIEW

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+        if add_background_features:
+            for feature_idx in range(n_features):
+                if feature_idx % 2 == 0:
+                    ax.axhspan(
+                        feature_idx - 0.5,
+                        feature_idx + 0.5,
+                        color="lightgray",
+                        alpha=0.1,
+                        zorder=0,
+                    )
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]:
```suggestion
            for feature_idx in range(0, n_features, 2):
                ax.axhspan(
                    feature_idx - 0.5,
                    feature_idx + 0.5,
                    color="lightgray",
                    alpha=0.1,
                    zorder=0,
                )
```

File: skore/src/skore/_sklearn/_plot/base.py [hunk 2]
```diff
@@ -118,11 +135,18 @@ def _style_params(self) -> list[str]:
...
+        self, *, policy: Literal["override", "update"] = "override", **kwargs: Any
+    ):
         """Set the style parameters for the display.
 
         Parameters
         ----------
+        policy : {"override", "update"}, default="override"
+            Policy to use when setting the style parameters. If "override", the style
+            parameters will be set to the values provided. If "update", the style
+            parameters will be updated with the values provided.
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/base.py]: If override, existing settings are set to the provided values. If update, existing settings are not changed; only settings that were previously unset are changed.

Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/base.py]: Yep I think that I'll take out those changes in another PR.

Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/base.py]: I opened the corresponding PR.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+                ax=self.ax_,
+                **plot_function_kwargs,
+            )
+            add_background_features = hue is not None and plot_function == sns.boxplot
+            self._decorate_matplotlib_axis(
+                ax=self.ax_,
+                add_background_features=add_background_features,
+                n_features=frame["feature"].nunique(),
+            )
+            self.ax_.set_title(f"{name}")
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]:
```suggestion
            self.ax_.set_title(name)
```

**Comment by auguste-probabl** [comment]
```python
from sklearn.datasets import load_iris
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from skore import EstimatorReport, train_test_split, ComparisonReport, CrossValidationReport
iris = load_iris(as_frame=True)
X, y = iris.data, iris.target
y = iris.target_names[y]
split_data = train_test_split(
    X=X, y=y, random_state=0, as_dict=True, shuffle=True
)
report_1 = CrossValidationReport(LogisticRegression(), X, y)
report_2 = CrossValidationReport(LogisticRegression(max_iter=500), X, y)
report = ComparisonReport([report_1, report_2])
display = report.feature_importance.coefficients()
display.plot()
```
image

The variance is very low which makes it difficult to see the colors.

**Comment by glemaitre** [comment]
> The variance is very low which makes it difficult to see the colors.

I'm not sure that we can do much. Here, the intercept value is dominant compared to the coefficients because we did not standardize the data. If you use a `StandardScaler` and the `LogisticRegression`, you will end-up with:

image

which is then manageable.

So I'm wondering if we should not revisit this issue when solving https://github.com/probabl-ai/skore/issues/1367 or with an option to whether or not to show the intercept.

Also I see that there is a small regression with the background. I think that the following is easier to read with the ordering of the legend and grouping.

image

**Comment by auguste-probabl** [comment]
> I'm not sure that we can do much. 

I think switching back to a horizontal bar plot with error bars would be a good first step

> Also I see that there is a small regression with the background.

Agreed, the shading does help!

**Comment by glemaitre** [comment]
OK, so I switch to seaborn to handle the columns as we discussed in https://github.com/probabl-ai/skore/issues/2190

I also implemented a mix of stripplot and boxplot for the problematic plot and allow to exclude the intercept if it is desired.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+        )
+
+    def _plot_matplotlib(
+        self,
+        *,
+        include_intercept: bool = True,
+        subplots_by: Literal["estimator", "label", "output"] | None = None,
+        barplot_kwargs: dict[str, Any] | None = None,
+        boxplot_kwargs: dict[str, Any] | None = None,
+        stripplot_kwargs: dict[str, Any] | None = None,
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py:227]:
```suggestion
        barplot_kwargs: dict[str, Any] = {},
        boxplot_kwargs: dict[str, Any] = {},
        stripplot_kwargs: dict[str, Any] = {},
```

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+                data=frame,
+                x="coefficients",
+                y="feature",
+                hue=hue,
+                col=col,
+                kind="strip",
+                dodge=True,
+                **stripplot_kwargs,
+            )
+            facet = facet.map_dataframe(
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]:
```suggestion
            ).map_dataframe(
```

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+                x="coefficients",
+                y="feature",
+                hue=hue,
+                dodge=True,
+                **boxplot_kwargs,
+            )
+            add_background_features = hue is not None
+
+        self.figure_, self.ax_ = facet.figure, facet.axes.squeeze()
+        # self.figure_.suptitle(name)
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: forgot to remove this line?

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py (hunk 4 above)
Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py:227]: We cannot do this one because a dictionary is mutable: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py:227]: Right.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+        stripplot_kwargs: dict[str, Any] | None = None,
+    ) -> None:
+        """Plot the coefficients for the different features.
+
+        Parameters
+        ----------
+        include_intercept : bool, default=True
+            Whether or not to include the intercept in the dataframe.
+
+        subplots_by : {"estimator", "label", "output"}, default=None
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: Let's choose between `subplots_by` and `subplot_by`. I prefer without the s (`subplot_by`) but its not a strong opinion.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+            The column to use for subplotting and dividing the coefficients into
+            subplots. If `None`, an automatic choice is made depending on the type of
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: IMO `None` and `"auto"` should be distinct choices for the `subplots_by`. There may be situations where we want to allow not subploting (`subplots_by=None`) but not have it be the default behaviour e.g. the default is `subplots_by="label"`.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 9]
```diff
@@ -0,0 +1,630 @@
...
+            facet = sns.catplot(
+                data=frame,
+                x="coefficients",
+                y="feature",
+                hue=hue,
+                col=col,
+                kind="bar",
+                **barplot_kwargs,
+            )
+            add_background_features = False
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: Why not have the shaded background in every case ? I think keeping consistent between reports is better.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py (hunk 6 above)
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: The suptitle is missing. Having it contain information on the data source and pos_label, when relevant, [like here](https://github.com/probabl-ai/skore/pull/2193/changes#diff-91d04a59777637c71fb2df63da34b82fa94a8b0a02790fa7acfbccf7095ee4d4R285-R287) would be nice I think.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py (hunk 7 above)
Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: yep let's go with singular to be aligned with `group_by` for instance.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py (hunk 8 above)
Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: Yes it is a good remark.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py (hunk 9 above)
Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: It does not look nice (I tried it). The bars already give an effect where you can easily parse the information.

Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: Indeed in binary classif/regression I agree it does not look good. I was trying in multiclass classif and there I think it makes sense, see:
image
image
So I suggest setting 
```add_background_features = hue is not None``` 
independently of the report type.

Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: also the axis is a bit squished in this example :thinking:

Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: I'll check if I can limit the number of ticks

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 10]
```diff
@@ -0,0 +1,645 @@
...
+            Keyword arguments to be passed to seaborn's :func:`seaborn.stripplot` for
+            rendering the coefficients with a :class:`~skore.ComparisonReport` of
+            :class:`~skore.CrossValidationReport`.
+        """
+        # help mypy to understand the following variable types
+        hue: str | None = None
+
+        # {"estimator", "label"} or {"estimator", "output"} or {"estimator"}
+        columns_to_groupby = self._get_columns_to_groupby(frame=frame)
+        if subplot_by not in ("auto", None) and subplot_by not in columns_to_groupby:
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py:502]: `subplot_by=None` should not be allowed in multiclass classification for ComparisonReports as it lead to unreadable plots:

Image

Image

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 11]
```diff
@@ -0,0 +1,645 @@
...
+            frame=frame,
+            report_type=report_type,
+            hue=hue,
+            col=col,
+            barplot_kwargs=barplot_kwargs,
+            boxplot_kwargs=boxplot_kwargs,
+            stripplot_kwargs=stripplot_kwargs,
+        )
+
+        self.figure_.suptitle(f"Coefficients of {estimator_name}")
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: In multiclass classif  `subplot_by` can be `"label"` but it is not reflected in the title, making it inconsistent with ComparisonReport.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 12]
```diff
@@ -0,0 +1,645 @@
...
+        """
+        # {"label"} or {"output"} or {}
+        columns_to_groupby = self._get_columns_to_groupby(frame=frame)
+        if subplot_by == "auto":
+            subplot_by = None
+
+        if subplot_by is not None and not len(columns_to_groupby):
+            raise ValueError(
+                "No columns to group by. `subplot_by` is expected to be None."
+            )
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py:414]:
```suggestion
        if subplot_by is not None and not len(columns_to_groupby):
            raise ValueError(
                "No columns to group by. `subplot_by` is expected to be None or auto."
            )
```

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 13]
```diff
@@ -0,0 +1,645 @@
...
+        elif subplot_by is not None and subplot_by not in columns_to_groupby:
+            raise ValueError(
+                f"Column {subplot_by} not found in the frame. It should be one "
+                f"of {', '.join(columns_to_groupby)}."
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]:
```suggestion
                f"of {', '.join(columns_to_groupby + ["auto", "None"])}."
```

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py [hunk 14]
```diff
@@ -0,0 +1,645 @@
...
+        """
+        # help mypy to understand the following variable types
+        hue: str | None = None
+
+        # {"estimator", "label"} or {"estimator", "output"} or {"estimator"}
+        columns_to_groupby = self._get_columns_to_groupby(frame=frame)
+        if subplot_by not in ("auto", None) and subplot_by not in columns_to_groupby:
+            raise ValueError(
+                f"Column {subplot_by} not found in the frame. It should be one "
+                f"of {', '.join(columns_to_groupby)}."
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py]: This error message should mention "auto", and None when in binary classification.

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py (hunk 12 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py:414]: "auto" with quotes

File: skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py (hunk 10 above)
Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/feature_importance/coefficients.py:502]: Done and added test

File: examples/use_cases/plot_feature_importance.py [hunk 15]
```diff
@@ -708,12 +712,13 @@ def add_y_true_pred(model_report, split):
...
-    title="Model weights",
-    xlabel="Coefficient",
-    ylabel="Feature",
+(
+    selectk_ridge_report.feature_importance.coefficients()
+    .frame()
+    .set_index("feature")
+    .sort_values(by="coefficients", key=abs, ascending=True)
+    .tail(15)
+    .plot.barh(title="Model weights", xlabel="Coefficient", ylabel="Feature")
```
Comment (GaetandeCast) [inline examples/use_cases/plot_feature_importance.py:721]: Just a thought: it would be nice to have a parameter that allows us to only plot the top or bottom n features to avoid such code. We could promote the use of such a parameter in a warning when the number of features is large.
It could be the subject of a follow-up PR.

Comment (glemaitre) [reply examples/use_cases/plot_feature_importance.py:721]: Yep, let's do in another PR but it is a good suggestion. It should not be the default because it depends on what type of analysis you are going to do with the magnitude.

**Comment by glemaitre** [comment]
OK so I addressed all reviews.

File: skore/src/skore/_sklearn/_plot/utils.py [hunk 16]
```diff
@@ -250,19 +251,29 @@ def _despine_matplotlib_axis(
     ----------
     ax : matplotlib.axes.Axes
         The matplotlib axis to despine.
+    axis_to_despine : tuple of str, default=("top", "right")
+        The axes to despine.
+    remove_ticks : bool, default=True
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/utils.py]:
```suggestion
    remove_ticks : bool, default=False
```

Line 243: `remove_ticks: bool = False,`

Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/utils.py]: Done
